File: sprint8/By lessons/lesson5code1/lesson5code1.go
N := len(s)
// Создадим массив π, состоящий из N нулей.
π := make([]int, N)

for i := 1; i < N; i++ {
    // i -- длина подстроки-префикса.
    substring := s[0:i]
    // Проверяем, перекрывается ли подстрока substring с собой по k символам.
    for k := i - 1; k >= 0; k-- {
        // Для этого сравним префикс и суффикс соответствующих длин.
        prefix := substring[0:k]
        suffix := substring[i-k:i]
        if prefix == suffix {
            // Запишем значение π-функции.
            π[i-1] = k
            // Дальше не проверяем -- k идёт в порядке уменьшения.
            break
        }
    }
}

File: sprint8/By lessons/lesson5code1/lesson5code1.js
const N = s.length;
const π = new Array(N).fill(0);

for (let i = 1; i < N; i++) {
  const substring = s.slice(0, i);
  for (let k = i - 1; k >= 0; k--) {
    const prefix = substring.slice(0, k);
    const suffix = substring.slice(i - k, i);
    if (prefix === suffix) {
      π[i - 1] = k;
      break;
    }
  }
}
